### PR TITLE
Fix stuttering when starting to scroll a Game List

### LINKED ISSUE #753: [GameList][Performance] Redunant Calls to updateInfoPanel() when scrolling by 1
I noticed significant stuttering on my system when scrolling on my game lists. I put the debugger on it, and found that it was spending most of it's time in `VideoGameListView::updateInfoPanel()` which makes sense as this is doing file access to the video, marquee, and images to be displayed... but what I found odd was that it was called 3 times for the same game (performing file accesses each time!) and when I only scrolled by 1 (which seems rather unnecessary). It was called once by the input for the `up` or `down` button, then it was called twice again for by the stop scrolling.

It calls from the `up` or `down` button from here:
https://github.com/RetroPie/EmulationStation/blob/c6bbd38d53d0d1a8929f01e32ebd4b72473c6422/es-app/src/components/TextListComponent.h#L283
Which then calls `listInput(delta)` then calls `scroll()`, when then calls `onCursorChanged()` which then calls the `updateInfoPanel()` later on.

The stop scrolling is from here:
https://github.com/RetroPie/EmulationStation/blob/c6bbd38d53d0d1a8929f01e32ebd4b72473c6422/es-app/src/components/TextListComponent.h#L313
This function is even more strange to me... It calls `listInput(0)` which will call the `onCursorChanged()` function calling `updateInfoPanel()`, but then it performs a direct call to `onCursorChanged()` right after finishing `listInput(0)`.

**VideoGameListView::updateInfoPanel()**
https://github.com/RetroPie/EmulationStation/blob/c6bbd38d53d0d1a8929f01e32ebd4b72473c6422/es-app/src/views/gamelist/VideoGameListView.cpp#L247

I tried to think to myself.. "How can I fix this?", but as I was reading through the lower portions of the code, but more I just began to see spaghetti 🍝 (or I just need to spend more time understanding it) and wasn't quite sure how to re-architect it in order to fix these redundant files calls without breaking anything else.

Anyways, I'm raising this issue just so it gets visibility from those who may have insights

### PR DESCRIPTION
**Pull request by XenuIsWatching**
Emulations station would stutter a lot when scorlling through game lists that had artwork with. This was due to it performing multiple reads of the screenshots, marquees, video, etc... For example, even if you were scrolling by 1, it would read these data 3 times. Each time it was calling the function `updateInfoPanel` which can kick of some File I/O which can be expensive (especially when you have the data on a NAS)

This gets it down to just 1 read. It also changes the behavior a little (for the better). There are three states in an input for scolling through a list: a rising edge where the button is pressed, a held high state where the button is held, and a falling edge where the button is released.

In the case where it was just press quickly, it was doing a read when the button was first pressed of the next item in the list. Then when it was released it would do another read of the next item in the list TWICE! The behavior now is to blank the info panel when the button is pressed, and then perform ONE read of the screenshot, etc. when the button is released.

Fixes #753

### CODE REVIEW

**Comment by XenuIsWatching** [comment]
I have this set to a draft because I want to run through some more test cases to make sure I didn't break anything else...

File: es-app/src/views/gamelist/DetailedGameListView.cpp [hunk 1]
```diff
@@ -212,15 +212,19 @@ void DetailedGameListView::initMDValues()
 
 void DetailedGameListView::updateInfoPanel()
 {
-	FileData* file = (mList.size() == 0 || mList.isScrolling()) ? NULL : mList.getSelected();
+	FileData* file = (mList.size() == 0 || (mList.getScrollingVelocity() != 0)) ? NULL : mList.getSelected();
```
Comment (pjft) [inline es-app/src/views/gamelist/DetailedGameListView.cpp:215]: Why this change? "isScrolling" seems to return a similar result - shouldn't we probably fix it at the isScrolling definition if there's an issue?

File: es-core/src/components/IList.h [hunk 2]
```diff
@@ -302,7 +298,7 @@ class IList : public GuiComponent
 			onScroll(absAmt);
 
 		mCursor = cursor;
-		onCursorChanged((mScrollTier > 0) ? CURSOR_SCROLLING : CURSOR_STOPPED);
+		onCursorChanged((amt != 0) ? CURSOR_SCROLLING : CURSOR_STOPPED);
```
Comment (pjft) [inline es-core/src/components/IList.h:301]: Tell us more about this change.

File: es-app/src/views/gamelist/DetailedGameListView.cpp (hunk 1 above)
Comment (XenuIsWatching) [reply es-app/src/views/gamelist/DetailedGameListView.cpp:215]: `isScrolling` also checks if ScrollTier is >0. What this means is that the user has held down the button and has now scrolled by at least 2 items. This is what was causing the stutter you begin to scroll, because it would load the the info.

I am considering add another function `isSingleScrolling` (or a better name)... (notice that mScrollTier is check to be == 0)
```c
bool isSingleScrolling(void)
{
    return (mScrollVelocity != 0 && mScrollTier == 0);
}
```

Right now the current implementation, will blank the info panel momentarily even if the user is just scrolling by one. Another option would be to have it just to do an instant transition (and not blank), but just not update the info panel retain what was there previously, and only update when the user has let go of the button.
It would only blank the info panel if the user has scrolled by at least 2.
Thoughts??

File: es-core/src/components/IList.h [hunk 3]
```diff
@@ -191,10 +191,6 @@ class IList : public GuiComponent
 	{
 		PowerSaver::setState(velocity == 0);
 
-		// generate an onCursorChanged event in the stopped state when the user lets go of the key
-		if(velocity == 0 && mScrollVelocity != 0)
-			onCursorChanged(CURSOR_STOPPED);
```
Comment (XenuIsWatching) [inline es-core/src/components/IList.h:196]: I want to make sure that removing this didn't break anything else.

The call to `stopScrolling()` would also call `onCursorChanged()` so that is where it would be called once, but believe there are other places that only call `listInput(0);`

File: es-core/src/components/IList.h (hunk 2 above)
Comment (XenuIsWatching) [reply es-core/src/components/IList.h:301]: I believe the intent of `mScrollTier` was to control and watch the speed of how much the user has scrolled. However mScrollTier is 0 when the user has just scrolled by 1, and only increments to 1 after the user has scrolled by 2. It is a better indicator that check that amt is not 0 to indicate that the user is not scrolling. As this function is also called when the user has stopped scrolling with amt == 0

Comment (pjft) [reply es-core/src/components/IList.h:301]: Thanks.

File: es-app/src/views/gamelist/DetailedGameListView.cpp (hunk 1 above)
Comment (pjft) [reply es-app/src/views/gamelist/DetailedGameListView.cpp:215]: What happens if you remove the mScrollTier check? If it needs to be there, I'd probably rather have isScrolling to receive an argument to distinguish whether you want to check mScrollTier or not as well.

**Comment by Gemba** [comment]
First: 
Thanks for documenting the variables. +1  Makes sense to me. However, I guess the wording needs some review (e.g. surplus "the"). My suggestion: Why not renaming `mScrollTier` to `mScrollGear` and `mScrollVelocity` to `mScrollStep` (as it is literally a fixed value and not a quotient over time)? This way `mScrollGearAccumulator` would almost explain itself.

Second: 
I found a small regression.

Large gamelist (> 1 Screen):
Single step: Fade out/fade in (differs from HEAD: no fading)
Page step: Fade out/fade in (differs from HEAD: no fading)

Short list (< 1 Screen):
Single step: Fade out/fade in (differs from HEAD: no fading)
Page step: No fading / hard skip (ok as in HEAD)

These are ok:
Continous scrolling / Dimmed with Initials: ok (same as in HEAD)
Roll-over (i.e. pushing up at first list item, pushing down on last list item): ok (same as in HEAD)

The introduced fadeout/fadein for single actions make the ES feel less snappy IMHO.

**Comment by XenuIsWatching** [comment]
> First:
> Thanks for documenting the variables. +1 Makes sense to me. However, I guess the wording needs some review (e.g. surplus "the"). My suggestion: Why not renaming `mScrollTier` to `mScrollGear` and `mScrollVelocity` to `mScrollStep` (as it is literally a fixed value and not a quotient over time)? This way `mScrollGearAccumulator` would almost explain itself.

`mScrollVelocity` is really the direction you are scrolling. '-1` for `up, `1` for `down`, and `0` for stop scrolling. I could be renaming variables all over the place for clairty :/ ... but I didn't want to go down that rabbit hole. That could be rather a separate pull request, as I want this to stay focused on the issue at hand rather than updating the code for easier maintance and readability.

> Second:
> I found a small regression in the DetailGamelistView.
> 
> Large gamelist (> 1 Screen):
> Single step: Fade out/fade in (differs from master: no fading)
> Page step: Fade out/fade in (differs from master: no fading)

Thanks for testing this out! 
About the single step, please see my comment below.
For the page step, I believe the stutter was so bad, you never saw the fading that was supposed to be happening...

> Short list (< 1 Screen):
> Single step: Fade out/fade in (differs from master: no fading)
> Page step: No fading / hard skip (ok as in master)

Again, Thanks for testing this out! 
Single step, same as above...
Page step, hmm... that's interesting that it was different behavior. I'll try to reproduce it, and investigate why that is happening.

> These are ok:
> Continous scrolling / Dimmed with Initials: ok (same as in master)
> Roll-over (i.e. pushing up at first list item, pushing down on last list item): ok (same as in master)

There is a bug I noticed in master with the Continuous scrolling, and then hitting the very end of the list. When it hits the very end of the list, It will load the info panel, even if the user hasn't let go of the button. But then once the user lets go of the button, it will reload the info panel again. You can notice this if you have a long description, and you hold the button until it begins scrolling after a brief moment. When you let go of it, the description will reset back to beginning indicateing it reloaded.

But this is very minor bug IMO...

> The introduced fadeout/fadein for single actions make the ES feel less snappy IMHO.

I thought this at first, and I did experiment with this using that `isSingleScrolling` function I mentioned earlier. What I did was just force it to retain the state of the info panel when the button was first pressed, and then blank once it is known that the user is holding it down (scrolling by at least 2). But IMHO, I actually felt the opposite and found that this implementation felt less snappier.
My thought went like this....

I press down a button, I see the next item selected... then I see the previous item still on the screen (albeit very briefly) wondering what's going on asking myself "Why am I still seeing old information on the screen, I already see the ?", and then I release the button, I finally see the update info after very brief moment again. All of this happens in less than quarter a second...

For the case I have implemented, my thought went like this...

I press down a button, I see the info panel fade out (and I think to myself it must be doing something and I feel like it's moving), and then I finally release the button, I see the update info after very brief moment again. Again, all of this happens in less than quarter a second...

This is just IMHO as well, but if there is consensus from the maintainers of this project, then I can easier make it perform the former action.

**Comment by XenuIsWatching** [comment]
I decided to go a simpler route which did no introduce blank in between single scrolling. All this does is record what it just wrote to the info panel, so that when it gets called again (for example, on a button release). it won't waste time updating the panel.

I found that in my system this function can take a while by adding in my own timers ( with `<chrono>`). I compared both my change and the base code. In the base code it calls it twice for button releases. I found that the first call (when the button is first pressed (rising edge), the time for this function can quite a long time under certain conditions in my system (50ms to 1000ms), but the subsequent calls that happened on the button release would take less (5ms to 35ms each). I can only speculate why the subsequent calls are so much shorter, but I would guess there is either some caching going on or my hard drive controller is doing some "magic". These subsequent calls are just rereading the same data and completely redundant and a waste of time.

Logs of my timing are below...
[retropie_infopanel_mod_time.txt](https://github.com/RetroPie/EmulationStation/files/6551491/retropie_infopanel_mod_time.txt)
[retropie_infopanel_base_time.txt](https://github.com/RetroPie/EmulationStation/files/6551492/retropie_infopanel_base_time.txt)

**Comment by Gemba** [comment]
I tried the suggested change on a installation with the ProfilingUtil (#716). 
As expected there was significant effect measurable (Rpi4 with UASP USB to SATA adapter):

```
es-app/src/views/gamelist/DetailedGameListView.cpp
Commit: 52c04d7
Message          	       Calls	  Total Time	    Avg Time	    Min Time	    Max Time	 Internal Total Time	   Internal Avg Time
updateInfoPanel()	         316	    9.973043	    0.031560	    0.000001	    0.108529	            9.973043	            0.031560

With this PR:
Message          	       Calls	  Total Time	    Avg Time	    Min Time	    Max Time	 Internal Total Time	   Internal Avg Time
updateInfoPanel()	         314	   10.098609	    0.032161	    0.000001	    0.104449	           10.098609	            0.032161
```


Could you pls try on your setup with the NAS? And: Are you using a wired or wireless connection to your NAS?

In a nutshell:
`cmake . -DPROFILING=on -DRPI=on -DGL=on`, `#include "utils/ProfilingUtil.h"` in object under test, then `ProfileScope("...");` in the method to profile. Start ES with `--debug`.
